src/testing/conditional_independence.py

```diff
@@ -0,0 +1,150 @@
+# -*- coding: utf-8 -*-
+"""
+Functions for performing permutation-based conditional independence tests.
+"""
+import numpy as np
+
+from sklearn.linear_model import LinearRegression
+from sklearn.metrics import r2_score
+
+import seaborn as sbn
+import matplotlib.pyplot as plt
+
+from tqdm import tqdm
+
+
+def _make_regressor(x_2d, y):
+    # Note, linear regression used instead of Random Forest Regression since
+    # the conditional independence test based on RandomForest regression did
+    # not pass the unit tests meant to assess whether the test statistic was
+    # a u-statistic.
+    regressor = LinearRegression()
+    regressor.fit(x_2d, y)
+    return regressor
+
+
+def computed_vs_obs_r2(df,
+                       x1_col,
+                       x2_col,
+                       z_col,
+                       mode_id_col,
+                       seed,
+                       num_permutations=100,
+                       progress=True):
+    # Set a random seed for reproducibility
+    if seed is not None:
+        np.random.seed(seed)
+
+    # Get the values to be used for testing
+    filter_array = (df[mode_id_col] == 1).values
```

Comment:
Need to change this so that no hard-coded filtering happens within the function.

The overall function should directly take the arrays to be used in the test rather than creating those arrays from the dataframe within the test.